File: CCC-2021/S3.cpp
#include <bits/stdc++.h>

using namespace std;

int N;
vector<int> P, W, D;

long long get_time(int pos) {
    long long seconds = 0;
    for (int i=0;i<N;i++) {
        long long distance = abs(P[i] - pos) - D[i];
        if (distance > 0)
            seconds += W[i] * distance;
    }
    return seconds;
}

int main()
{
    ios_base::sync_with_stdio(0); cin.tie(0); cout.tie(0);
    int l = INT_MAX, r = INT_MIN, mid;
    long long shortest_time;

    cin >> N;
    P.resize(N);
    W.resize(N);
    D.resize(N);

    for (int i=0;i<N;i++) {
        cin >> P[i] >> W[i] >> D[i];
        if (P[i] < l)
            l = P[i];
        if (P[i] > r)
            r = P[i];
    }

    while (l <= r) {
        mid = l + (r - l) / 2;

        long long time = get_time(mid);
        long long timePlusOne = get_time(mid+1);
        long long timeMinusOne = get_time(mid-1);

        if ((time <= timePlusOne) && (time <= timeMinusOne)) {
            shortest_time = time;
            break;
        } else if (time < timePlusOne) {
            r = mid - 1;
        } else {
            l = mid + 1;
        }
    }

    cout << shortest_time << "\n";
    return 0;
};

File: CCC-2021/S3.py
import sys

def getTime(position):
    secs = 0
    for i in range(0, n):
        distance = abs(p[i] - position) - d[i]
        if distance > 0:
            secs += w[i] * distance
    return secs

p, w, d = [], [], []
n, l, r = int(input()), sys.maxsize, -sys.maxsize - 1

for i in range(0, n):
    values = input().split(" ")
    p.append(int(values[0]))
    w.append(int(values[1]))
    d.append(int(values[2]))

    l = min(l, p[i])
    r = max(r, p[i])

while (l <= r):
    mid = round(l + (r - l) / 2)

    time = getTime(mid) 
    timeplusOne = getTime(mid+1)
    timeMinusOne = getTime(mid-1)

    if (time <= timeplusOne) and (time <= timeMinusOne):
        shortestTime = time 
        break 
    elif time < timeplusOne:
        r = mid - 1
    else: 
        l = mid + 1

print(shortestTime)


